datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -370,7 +391,12 @@ struct ExprIdentifierVisitor<'a> {
     expr_set: &'a mut ExprSet,
     /// series number (usize) and identifier.
     id_array: &'a mut Vec<(usize, Identifier)>,
-    data_type: DataType,
+    /// input schema for the node that we're optimizing, so we can determine the correct datatype
+    /// for each subexpression
+    input_schema: DFSchemaRef,
+    /// all schemas in the logical plan, as a fall back if we cannot resolve an expression type
+    /// from the input schema alone
+    all_schemas: Vec<DFSchemaRef>,
```

Comment:
I don't understand in what cases we wouldn't be able to resolve an expr type from the input schema alone.

The only case I can think of is when the plan node has more than one input (e.g. a Join or a Union) -- but thus I would expect that we always resolve the type of the expressions using the input schema

Reply:
While Joins and Unions are not (yet?) handled by this optimization rule, I think that even in these cases we should be able to construct one consolidated schema that is used to resolve the expression type - otherwise the expression probably is invalid in the first place.

Right now the fall-back logic just randomly merges all possible schemas into one - there's no guarantee the resulting, merged schema will be any good for resolving the expression at hand. That's especially the case if the logical plan involves lots of aliasing - there may be many fields from different nodes that have the same name, e.g. `a`, but different data types; the merged schema will have only one `a` column, though, the first that we encounter while merging schemas.

Reply:
I am glad I was not the only one confused :)

Reply:
This random merge used to only exist in the `Filter` plan and was fixed once in https://github.com/apache/arrow-datafusion/pull/1925 . Ashamedly, I cannot recall why `Filter` is unique :disappointed_relieved: I would like to explain it as my mistake, sorry for that >_<

I think it is doable to remove the else branch as it already takes no effect after #1925

Reply:
I wouldn't worry -- there have been many changes in DataFusion and it was (and still is) a fast changing codebase -- the day I write perfect code without any errors will be the day I hold other people to the same standard 😆